File: packages/python/src/tokenomics/experiment.py
from __future__ import annotations

import hashlib
import json
from typing import Any

TREATMENT_FIELDS = (
    "model_version",
    "reasoning_effort",
    "system_prompt_hash",
    "skills_hash",
    "context_policy",
    "tool_schema_hash",
    "temperature",
)


def treatment_hash(**fields: Any) -> str:
    """Stable 16-hex fingerprint for an experiment treatment.

    The canonical payload always contains every known treatment field, using null
    for omitted values. This deliberately matches the z0int receipt semantics.
    """
    payload = {key: fields.get(key) for key in TREATMENT_FIELDS}
    blob = json.dumps(payload, sort_keys=True, separators=(",", ":"), ensure_ascii=False).encode("utf-8")
    return hashlib.sha256(blob).hexdigest()[:16]

File: packages/typescript/src/experiment.ts
import { createHash } from "node:crypto";

const FIELDS = [
  "model_version", "reasoning_effort", "system_prompt_hash", "skills_hash",
  "context_policy", "tool_schema_hash", "temperature"
] as const;

export type TreatmentFields = Partial<Record<(typeof FIELDS)[number], string | number | null>>;

export function treatmentHash(fields: TreatmentFields = {}): string {
  const payload: Record<string, string | number | null> = {};
  for (const key of [...FIELDS].sort()) payload[key] = fields[key] ?? null;
  return createHash("sha256").update(JSON.stringify(payload)).digest("hex").slice(0, 16);
}
